balancer/endpointsharding/endpointsharding_test.go

```diff
@@ -158,25 +159,62 @@ func (s) TestEndpointShardingBasic(t *testing.T) {
 	}
 }
 
-// TestEndpointShardingBalancerAPI tests that when endpoint sharding gets a
-// Client Conn update with a duplicate address across endpoints the
-// UpdateClientConnState operation errors.
-func (s) TestEndpointShardingBalancerAPI(t *testing.T) {
-	es := NewBalancer(nil, balancer.BuildOptions{})
-	addr := resolver.Address{Addr: "addr1"}
-	err := es.UpdateClientConnState(balancer.ClientConnState{
-		ResolverState: resolver.State{
-			Endpoints: []resolver.Endpoint{
-				{Addresses: []resolver.Address{addr}},
-				{Addresses: []resolver.Address{addr}},
+// TestEndpointShardingBalancerResolverAddresses tests different scenarios of
+// resolver addresses being updated to the endpoint sharding balancer. Some
+// should cause errors and some shouldn't.
+func (s) TestEndpointShardingBalancerResolverAddresses(t *testing.T) {
+	addr1 := resolver.Address{Addr: "addr1"}
+	addr2 := resolver.Address{Addr: "addr2"}
+	addr3 := resolver.Address{Addr: "addr3"}
+	addr4 := resolver.Address{Addr: "addr4"}
+	tests := []struct {
+		name      string
+		endpoints []resolver.Endpoint
+		wantErr   bool
+	}{
+		{
+			name: "duplicate-address-across-endpoints",
+			endpoints: []resolver.Endpoint{
+				{Addresses: []resolver.Address{addr1}},
+				{Addresses: []resolver.Address{addr1}},
 			},
+			wantErr: true,
+		},
+		{
+			name: "duplicate-address-across-endpoints-plural-addresses",
+			endpoints: []resolver.Endpoint{
+				{Addresses: []resolver.Address{addr1, addr2, addr3}},
+				{Addresses: []resolver.Address{addr3, addr4}},
+			},
+			wantErr: true,
+		},
+		{
+			name: "no-shared-addresses",
+			endpoints: []resolver.Endpoint{
+				{Addresses: []resolver.Address{addr1, addr2}},
+				{Addresses: []resolver.Address{addr3, addr4}},
+			},
+			wantErr: false,
+		},
+		{
+			name: "endpoint-with-no-addresses",
+			endpoints: []resolver.Endpoint{
+				{Addresses: []resolver.Address{addr1, addr2}},
+				{Addresses: []resolver.Address{}},
+			},
+			wantErr: true,
 		},
```

Comment:
Please add a couple more:

- no error with addr1+2 and addr3+4 in separate endpoints -- you should generally always have a "pass" test.
- error where endpoint1 has the same addr multiple times

Reply:
Discussed offline; had the first case. Added the second case.